Intro example at /hello

nalaka opened the pull request:
Introductory example app demonstrating the use of the **JavaScript SDK**
for **Cognite Fusion** with **Gearbox** React components.

- Single page app bootstrapped with Create React App
- Uses `CogniteClient` from the JS SDK and `ClientSDKProvider` from Gearbox
- Uses `AssetTree` and `AssetMeta` components to display data
- Uses `Layout`, `Row`, and `Col` components from Ant Design

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

nalaka commented:
Thanks @f1cognite, when I tried the [preview URL](https://pr-62.javascript-getting-started.preview.cogniteapp.com/hello/), I got few errors like below:

`Uncaught SyntaxError: Unexpected token '<' `

Maybe an issue when doing the production build?

Works ok with a local production build served over HTTPS though.

```sh
cd hello
npm install
npm run build

cd build
https-serve
```
<img width="278" alt="ex-hello-prod-local" src="https://user-images.githubusercontent.com/437578/65059196-2e329e80-d993-11e9-85d0-284422f65b2e.png">

f1cognite commented on hello/.gitignore:
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
This file can be removed (we have a .gitignore file in the root folder)

f1cognite commented on hello/README.md:
````diff
@@ -0,0 +1,75 @@
+```shell
+npm install
+
+HTTPS=true npm start
````
Could we add the `HTTPS=true` statement inside the `start`-script in package.json? See here: https://github.com/cognitedata/javascript-getting-started/blob/master/infographic-app/package.json#L17

f1cognite commented on hello/README.md:
````diff
@@ -0,0 +1,75 @@
+}
+```
+
+The `CogniteClient` component nicely packages the complexity of authentication
````
Great!

f1cognite commented on hello/package-lock.json:
```diff
@@ -0,0 +1,15017 @@
+{
```
Can you remove this file and use yarn instead? So we get a `yarn.lock` file instead of a `package-lock.json` file? That will be consistent with the other examples

f1cognite commented on hello/src/App.js:
```diff
@@ -0,0 +1,45 @@
+    });
+
+    client.loginWithOAuth({ project: PROJECT_ID });
+    client.authenticate().then(() => {
```
Use `await` instead? Since the function is using `async`

f1cognite commented on hello/src/App.js:
```diff
@@ -0,0 +1,45 @@
+import "antd/dist/antd.css";
+
+const APP_ID = "hello-cdf-gearboxjs";
+const PROJECT_ID = "publicdata";
```
Challenge to you 😄 : Use the TenantSelector so the user can choose their own project. Take a look at the "charting-app" example

We can also drop this to make the example app even simpler. What do you think @polomani ?

polomani replied:
Since we already have https://github.com/cognitedata/javascript-getting-started/tree/master/asset-explorer-app which uses TenantSelector, AssetMeta, AssetSearch and AssetTree  I would rather go for a simpler app.
Consisting of just:
- List of 5/10 assets which you can choose (fetched with just an sdk, not a gearbox component, maybe filtered Idk)
- AssetMeta for a selected one
- One page app, so we don't have back button
- No tenant selector for simplicity, public data is by default

Thoughts @f1cognite ?

f1cognite replied:
Sounds good!

polomani replied:
We talked with @f1cognite a decided that we need to simplify this app a little bit (if you agree of course) 

The reason for that is that we already have https://github.com/cognitedata/javascript-getting-started/tree/master/asset-explorer-app which uses TenantSelector, AssetMeta, AssetSearch and AssetTree.
So we think we should go for a tiny app consisting of just:
- AssetTree
- AssetMeta for a selected one
- One page app, so we don't have back button
- No tenant selector for simplicity, public data is hardcoded as it is right now

This I believe would be more useful for beginners and different from other examples.

polomani replied:
For you it's mostly just removing a button and make a different layout. :)

f1cognite commented on hello/src/Explorer.js:
```diff
@@ -0,0 +1,69 @@
+import React from "react";
+import { AssetTree, AssetMeta } from "@cognite/gearbox";
+import { Button } from "antd";
```
Put `Button` in the import below

f1cognite commented on hello/src/Explorer.js:
```diff
@@ -0,0 +1,69 @@
+          {this.state.isTreeVisible && (
+            <AssetTree
+              onSelect={asset => {
+                console.log(`AssetTree: selected ${asset.key}`);
```
Remove this

polomani commented on hello/public/robots.txt:
```diff
@@ -0,0 +1,2 @@
+# https://www.robotstxt.org/robotstxt.html
```
I don't think we need this file.

Also please, rename the app and it's folder to "hello-gearbox" or something more explicit.
And move screenshots to a separate folder

nalaka commented:
Thanks for the detailed review @f1cognite & @polomani - incorporated all of it.

- Set env var `HTTPS` to `true` inside `package.json` itself
- Use Yarn instead of NPM for consistency with other examples
- Use `await` instead of `then` in `App > componentDidMount`
- Cleaned **antd** imports in **Explorer**
- Removed unnecessary log in `Explorer > AssetTree > onSelect`
- Removed unnecessary local `.gitignore` file
- Removed `public/robots.txt`
- Moved screenshots to `res` (resources) folder
- Renamed app and folder to **hello-gearbox**
- Added to the example list on top level README
- Back button moved to top-left corner for better discoverability

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

f1cognite commented:
@nalaka I see that the PR server still fails: https://pr-62.javascript-getting-started.preview.cogniteapp.com/hello-gearbox/

You need to add the following line to `package.json`: `"homepage: "."` as seen here: https://github.com/cognitedata/javascript-getting-started/blob/master/charting-app/package.json#L4

Great job! 🚀

polomani commented on hello-gearbox/README.md:
```diff
@@ -0,0 +1,75 @@
+# Hello Gearbox
+
+> Introductory example app demonstrating how to access the `Cognite Data Fusion Platform` with the `JavaScript SDK` and the `Cognite Gearbox` component library.
```
links to all of them would be appreciated :)

polomani replied:
you have them underneath, so just move them here (usually we put the link first time we mention something, after that it's not needed anymore)

polomani commented on hello-gearbox/README.md:
```diff
@@ -0,0 +1,75 @@
+
+> Introductory example app demonstrating how to access the `Cognite Data Fusion Platform` with the `JavaScript SDK` and the `Cognite Gearbox` component library.
+
+Run with:
```
```suggestion
Install npm modules:

`yarn` or `npm install`

Start application:

`yarn start`or `npm start`
```

polomani commented on hello-gearbox/README.md:
````diff
@@ -0,0 +1,75 @@
+```shell
+npm install
+
+HTTPS=true npm start
````
and then this can be removed

polomani commented on hello-gearbox/README.md:
```diff
@@ -0,0 +1,75 @@
+# Hello Gearbox
+
+> Introductory example app demonstrating how to access the `Cognite Data Fusion Platform` with the `JavaScript SDK` and the `Cognite Gearbox` component library.
```
```suggestion
Introductory example app demonstrating how to access the `Cognite Data Fusion Platform` with the `JavaScript SDK` and the `Cognite Gearbox` component library.
```

nalaka replied:
Fixed

nalaka commented:
- Added `homepage` field to `package.json`
- Moved links up to the first occurrence of the reference
- Fixed run instructions
- Updated componentDidMount code excerpt
- Updated the screenshot of AssetMeta component

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

nalaka commented:
Preview works - thanks for the tip @f1cognite

nalaka commented:
I think this last commit addresses the comment by @polomani about making the example be just a single page and getting rid of the back button.

Earlier, when the back button was tapped, the AssetTree got re-rendered.
This caused the selected asset to be reset. So if you selected an asset somewhere deep in the tree, view it, and  tap the back button, the tree would be collapsed. It was like the AssetTree and AssetMeta were in two different pages.

polomani commented on hello-gearbox/README.md:
```diff
@@ -0,0 +1,78 @@
+# Hello Gearbox
+
+Introductory example app demonstrating how to access the Cognite [Data Fusion Platform](https://cognite.com/products/cognite-data-fusion) with the [JavaScript SDK](https://github.com/cognitedata/cognitesdk-js) and the [Gearbox](https://github.com/cognitedata/gearbox.js) component library.
```
```suggestion
Introductory example app demonstrating how to access the Cognite [Data Fusion Platform](https://docs.cognite.com/) with the [JavaScript SDK](https://www.npmjs.com/package/@cognite/sdk) and the [Gearbox](https://www.npmjs.com/package/@cognite/gearbox) component library.
```

polomani commented on hello-gearbox/README.md:
```diff
@@ -0,0 +1,78 @@
+# Hello Gearbox
+
+Introductory example app demonstrating how to access the Cognite [Data Fusion Platform](https://docs.cognite.com/) with the [JavaScript SDK](https://www.npmjs.com/package/@cognite/sdk) and the [Gearbox](https://www.npmjs.com/package/@cognite/gearbox) component library.
```
```suggestion
Introductory example app demonstrating how to access the [Cognite Data Fusion](https://docs.cognite.com/) with the [JavaScript SDK](https://www.npmjs.com/package/@cognite/sdk) and the [Gearbox](https://www.npmjs.com/package/@cognite/gearbox) component library.
```

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

f1cognite commented:
This apps can now be previewed at the following URL: https://pr-62.javascript-getting-started.preview.cogniteapp.com

polomani commented:
thanks for the PR @nalaka

nalaka commented:
@polomani - there is a 404 error at the demo page:
https://cognitedata.github.io/javascript-getting-started/hello-gearbox

nalaka commented:
Now ok- thanks @polomani & @f1cognite for the reviews and guidance!
